src/fetch/transfer_file.py

```diff
@@ -1,51 +1,119 @@
+from __future__ import annotations
+
 import argparse
 from dataclasses import dataclass
 from datetime import datetime
+from enum import Enum
 from typing import Optional
 
 from src.dune_analytics import DuneAnalytics, QueryParameter
+from src.fetch.period_slippage import get_period_slippage, SolverSlippage
 from src.file_io import File, write_to_csv
 from src.models import Network, Address
+from src.utils.dataset import index_by
+
+
+class TokenType(Enum):
+    """
+    Classifications of CSV Airdrop Transfer Types
+    """
+    NATIVE = 'native'
+    ERC20 = 'erc20'
+
+    # Technically the app also supports NFT transfers, but this is irrelevant here
+    # NFT = 'nft'
+
+    @classmethod
+    def from_str(cls, type_str: str) -> TokenType:
+        """Constructs Enum variant from string (case-insensitive)"""
+        try:
+            return cls[type_str.upper()]
+        except KeyError as err:
+            raise ValueError(f"No TransferType {type_str}!") from err
+
+    def __str__(self):
+        return self.value
 
 
 @dataclass
 class Transfer:
     """Total amount reimbursed for accounting period"""
-    token_type: str
+    token_type: TokenType
+    # Safe airdrop uses null address for native asset transfers
     token_address: Optional[Address]
     receiver: Address
     # safe-airdrop uses float amounts!
     amount: float
 
-    def __init__(self, token_type, token_address, receiver, amount):
-        self.token_type = token_type
-        self.token_address = Address(token_address) if token_address else None
-        self.receiver = Address(receiver)
-        self.amount = float(amount)
+    @classmethod
+    def from_dict(cls, obj: dict) -> Transfer:
+        """Converts Dune data dict to object with types"""
+        token_type = TokenType.from_str(obj['token_type'])
+        token_address = obj['token_address']
+        if token_type == TokenType.NATIVE and token_address is not None:
+            raise ValueError("Native transfers must have null token_address")
+        if token_type == TokenType.ERC20 and token_address is None:
+            raise ValueError("ERC20 transfers must have valid token_address")
+
+        return cls(
+            token_type=token_type,
+            token_address=Address(
+                token_address) if token_type != TokenType.NATIVE else None,
+            receiver=Address(obj['receiver']),
+            amount=float(obj['amount'])
+        )
+
+    def add_slippage(self, slippage: Optional[SolverSlippage]):
+        if slippage is None:
+            return
+        assert self.receiver == slippage.solver_address
+        adjustment = slippage.amount_wei / 10 ** 18
+        print(
+            f"Adjusting {self.receiver} transfer by {adjustment:.5f} (slippage)"
```

Comment:
Can we also add here the solver name? In the final print, I would love to see which solver get the heavies slashing

Reply:
Yes I think I may have already added this locally. Just forgot to push.

Reply:
done.